apps/nsq_to_nsq/nsq_to_nsq.go

```diff
@@ -349,37 +402,45 @@ func main() {
 		hostPool = hostpool.NewEpsilonGreedy(destNsqdTCPAddrs, 0, &hostpool.LinearEpsilonValueCalculator{})
 	}
 
-	handler := &PublishHandler{
-		addresses:        destNsqdTCPAddrs,
-		producers:        producers,
-		mode:             selectedMode,
-		hostPool:         hostPool,
-		respChan:         make(chan *nsq.ProducerTransaction, len(destNsqdTCPAddrs)),
-		perAddressStatus: perAddressStatus,
-		timermetrics:     timer_metrics.NewTimerMetrics(*statusEvery, "[aggregate]:"),
-	}
-	consumer.AddConcurrentHandlers(handler, len(destNsqdTCPAddrs))
+	consumerList := initConsumerAndHandler(cCfg, producers, selectedMode, hostPool, perAddressStatus)
 
-	for i := 0; i < len(destNsqdTCPAddrs); i++ {
-		go handler.responder()
+	for _, consumer := range consumerList {
+		err := consumer.ConnectToNSQDs(nsqdTCPAddrs)
+		if err != nil {
+			log.Fatal(err)
+		}
 	}
 
-	err = consumer.ConnectToNSQDs(nsqdTCPAddrs)
-	if err != nil {
-		log.Fatal(err)
+	for _, consumer := range consumerList {
+		err := consumer.ConnectToNSQLookupds(lookupdHTTPAddrs)
+		if err != nil {
+			log.Fatal(err)
+		}
 	}
 
-	err = consumer.ConnectToNSQLookupds(lookupdHTTPAddrs)
-	if err != nil {
-		log.Fatal(err)
+	aggregateStopChan := make(chan bool)
+
+	for _, consumer := range consumerList {
+		go func(consumer *nsq.Consumer, aggregateStopChan chan<- bool) {
```

Comment:
Right - you want to wait for all the consumers, but your current `aggregateStopChan` strategy only waits for one of them.

The previous code waited for either a message on `termChan`, _or_ for the consumer to stop, in case it happened "on its own", before any message on `termChan`. I think that for reasonably recent versions of go-nsq, the consumer won't stop before being told to stop explicitly, it will keep trying to reconnect to nsqlookupd or nsqd. So I think you can just wait for `termChan` first, then `close()` all consumers, then wait for them all to stop.

I think that if that is all true, and that strategy is followed, then the `sync.WaitGroup` is not really needed. I think the `sync.WaitGroup` would have just replaced the `aggregateStopChan`, enabling a convenient reliable way to wait for all consumers instead of just one. You would still need a goroutine for each consumer, waiting on the `StopChan` and then calling `wg.Done()`. But I think you can skip that goroutine-per-consumer, and just loop through all consumers again after calling `.Stop()` on all of them, making sure each is done.